File: bot.py
require('dotenv').config();
const { Bot, InlineKeyboard } = require('grammy');
const axios = require('axios');

// ENV Configuration
const BOT_TOKEN = process.env.BOT_TOKEN;
const SMM_API_URL = process.env.SMM_API_URL || "https://smmlite.com/api/v2";
const SMM_API_KEY = process.env.SMM_API_KEY;
const SUPPORT_USERNAME = process.env.SUPPORT_USERNAME || "YourSupportUsername";
const UPI_ID = process.env.UPI_ID || "your-vpa@ybl";
const USDT_ADDRESS = process.env.USDT_ADDRESS || "TYxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx";
const USD_TO_INR_RATE = 95.0;

// Mock database (In-memory)
const USER_DATABASE = {};

// Mock SERVICES_MASTER_DATA (Python से SERVICES_MASTER_DATA को यहाँ परिभाषित करें)
const SERVICES_MASTER_DATA = {
    // उदाहरण के लिए:
    // "5153": { name: "Telegram Reactions", rate: 0.12, type: "tg_post" }
};

const bot = new Bot(BOT_TOKEN);

// Regex
const TG_POST_RE = /https:\/\/t\.me\/([A-Za-z0-9_]+)\/(\d+)\/?/;
const TG_CHANNEL_RE = /https:\/\/t\.me\/([A-Za-z0-9_]+)\/?/;
const IG_POST_RE = /https:\/\/(www\.)?instagram\.com\/(?:p|reel|reels|tv)\/([A-Za-z0-9_\-]+)\/?/;
const IG_PROFILE_RE = /https:\/\/(www\.)?instagram\.com\/([A-Za-z0-9_\.]+)\/?/;

// Helpers
function getOrCreateUser(userId) {
    if (!USER_DATABASE[userId]) {
        USER_DATABASE[userId] = {
            balance_usd: 0.0,
            spent_usd: 0.0,
            orders_count: 0,
            channels: [],
            history: [],
            order_details: {},
            currency: "INR",
            pending_service: null,
            pending_qty: null,
            pending_cost_usd: null
        };
    }
    return USER_DATABASE[userId];
}

function formatMoney(usd, pref) {
    if (pref === "INR") {
        return `₹${(usd * USD_TO_INR_RATE).toFixed(2)}`;
    }
    return `$${usd.toFixed(2)}`;
}

function validateLink(link, type) {
    if (type === "tg_post") return TG_POST_RE.test(link);
    if (type === "tg_channel") return TG_CHANNEL_RE.test(link);
    if (type === "ig_post") return IG_POST_RE.test(link);
    if (type === "ig_profile") return IG_PROFILE_RE.test(link);
    return true;
}

// Inline Keyboards Creators
function getMainAddFundsKeyboard() {
    return new InlineKeyboard()
        .text("₹100", "amt_100").text("₹200", "amt_200").row()
        .text("₹500", "amt_500").text("₹1000", "amt_1000").row()
        .text("₹2000", "amt_2000").text("₹5000", "amt_5000").row()
        .text("⬅️ Back", "back_to_menu");
}

// Handlers
bot.callbackQuery("toggle_currency", async (ctx) => {
    const user = getOrCreateUser(ctx.from.id);
    user.currency = user.currency === "USD" ? "INR" : "USD";
    await ctx.answerCallbackQuery({ text: `✔️ Currency set to ${user.currency}` });
    // यहाँ आप चाहें तो मेनू प्रोफाइल दोबारा रेंडर कर सकते हैं
});

bot.callbackQuery("main_add_funds", async (ctx) => {
    await ctx.editMessageText(
        `💳 *Add Funds / डिपाजिट Fund:*\n\nनीचे दिए गए अमाउंट्स में से कोई एक अमाउंट चुनें:`,
        { reply_markup: getMainAddFundsKeyboard(), parse_mode: "Markdown" }
    );
});

bot.callbackQuery(/^amt_\d+$/, async (ctx) => {
    const user = getOrCreateUser(ctx.from.id);
    const amtInr = parseInt(ctx.callbackQuery.data.split("_")[1]);
    const amtUsd = amtInr / USD_TO_INR_RATE;

    const keyboard = new InlineKeyboard()
        .text("✅ मैंने पेमेंट कर दी है", `paid_${amtInr}`).row()
        .text("⬅️ Back", "main_add_funds");

    const text = `💵 *Pay:* ₹${amtInr} (~$${amtUsd.toFixed(2)}) USD\n\n` +
                 `📍 *UPI ID:* \`${UPI_ID}\`\n` +
                 `🪙 *USDT Address:* \`${USDT_ADDRESS}\`\n\n` +
                 `⚠️ Pay karke screenshot support par bhejein.`;

    await ctx.editMessageText(text, { reply_markup: keyboard, parse_mode: "Markdown" });
});

bot.callbackQuery(/^paid_\d+$/, async (ctx) => {
    const user = getOrCreateUser(ctx.from.id);
    const amtInr = parseInt(ctx.callbackQuery.data.split("_")[1]);
    const amtUsd = amtInr / USD_TO_INR_RATE;

    user.history.push({
        type: "deposit",
        amount_inr: amtInr,
        amount_usd: amtUsd,
        status: "Pending ⏳"
    });

    const keyboard = new InlineKeyboard()
        .url("👤 Contact Support", `https://t.me{SUPPORT_USERNAME}`).row()
        .text("⬅️ Menu", "back_to_menu");

    await ctx.editMessageText(
        `💌 *Request Sent!*\n\nTeam verify karke balance add karegi.\nScreenshot support par bhejein.`,
        { reply_markup: keyboard }
    );
});

bot.callbackQuery("main_channels", async (ctx) => {
    const user = getOrCreateUser(ctx.from.id);
    const botInfo = await ctx.api.getMe();
    const promoteUrl = `https://t.me{botInfo.username}?startchannel=true&admin=post_messages+edit_messages+delete_messages+invite_users`;

    let text = `📢 *My Channels:*\n\n`;
    if (!user.channels || user.channels.length === 0) {
        text += `❌ Koi channel linked nahi hai.\n`;
    } else {
        user.channels.forEach((ch, idx) => {
            text += `${idx + 1}. ${ch.title} (Active ✅)\n`;
        });
    }

    const keyboard = new InlineKeyboard()
        .url("➕ Promote Bot as Admin", promoteUrl).row()
        .text("⬅️ Back", "back_to_menu");

    await ctx.editMessageText(text, { reply_markup: keyboard, parse_mode: "Markdown" });
});

bot.callbackQuery("main_support", async (ctx) => {
    const keyboard = new InlineKeyboard()
        .url("💬 Chat with Support", `https://t.me{SUPPORT_USERNAME}`).row()
        .text("⬅️ Back", "back_to_menu");

    await ctx.editMessageText(
        `📞 *Support*\n\nKisi bhi problem ke liye yahan contact karein:`,
        { reply_markup: keyboard, parse_mode: "Markdown" }
    );
});

bot.callbackQuery("main_promo", async (ctx) => {
    const keyboard = new InlineKeyboard().text("⬅️ Back", "back_to_menu");
    await ctx.editMessageText(
        `🎁 *Promotions*\n\n✨ Welcome Bonus: Coming soon!\n🤝 Refer & Earn: Coming soon!`,
        { reply_markup: keyboard, parse_mode: "Markdown" }
    );
});

bot.callbackQuery("main_orders", async (ctx) => {
    const user = getOrCreateUser(ctx.from.id);
    const keyboard = new InlineKeyboard().text("⬅️ Back", "back_to_menu");

    const orders = user.history.filter(h => h.order_id);

    if (orders.length === 0) {
        await ctx.editMessageText(`📦 *My Orders*\n\nAbhi tak koi order nahi hai.`, { reply_markup: keyboard, parse_mode: "Markdown" });
        return;
    }

    let text = `📦 *My Orders (Last 10):*\n\n`;
    const lastOrders = orders.slice(-10);
    for (const o of lastOrders) {
        text += `🆔 OID: \`${o.order_id}\`\n` +
                `🛠️ Service: ${o.service_id}\n` +
                `📊 Qty: ${o.qty}\n` +
                `💰 Cost: ${formatMoney(o.cost_usd, user.currency)} | Status: ${o.status}\n\n`;
    }

    await ctx.editMessageText(text, { reply_markup: keyboard, parse_mode: "Markdown" });
});

bot.callbackQuery("main_services", async (ctx) => {
    const keyboard = new InlineKeyboard()
        .text("🔹 TELEGRAM", "platform_telegram")
        .text("🔸 INSTAGRAM", "platform_instagram").row()
        .text("🔹 FACEBOOK", "platform_facebook")
        .text("🔸 YOUTUBE", "platform_youtube").row()
        .text("⬅️ Back", "back_to_menu");

    await ctx.editMessageText(
        `🛠️ *Select Platform / प्लेटफार्म चुनें:*\n\nAap kiski services dekhna chahte hain?`,
        { reply_markup: keyboard, parse_mode: "Markdown" }
    );
});

bot.callbackQuery("platform_telegram", async (ctx) => {
    const user = getOrCreateUser(ctx.from.id);
    const pref = user.currency;
    const r = (usd) => pref === "INR" ? `₹${(usd * USD_TO_INR_RATE).toFixed(2)}` : `$${usd.toFixed(2)}`;

    let text = `🔹 *TELEGRAM SERVICES* [Currency: ${pref}]\n\n` +
               `💬 *TELEGRAM REACTIONS*\n` +
               `▫️ /5153 - telegram like (👍) [instant] - ${r(0.12)} per 1000\n` +
               `▫️ /5160 - telegram like (👍❤️🔥🥰) [instant] - ${r(0.15)} per 1000\n` +
               `▫️ /5161 - telegram like (❤️🔥👏🤩🎉🥰👍) [instant] - ${r(0.10)} per 1000\n` +
               `▫️ /5162 - telegram like (🔥) [instant] - ${r(0.15)} per 1000\n` +
               `▫️ /5163 - telegram like (❤️) [instant] - ${r(0.15)} per 1000\n` +
               `▫️ /5164 - telegram like (👏) [instant] - ${r(0.15)} per 1000\n` +
               `▫️ /5165 - telegram like (🤩) [instant] - ${r(0.15)} per 1000\n\n` +
               `👀 *TELEGRAM POST VIEWS*\n` +
               `▫️ /1512 - telegram post views [Last 1 post] [SUPERFAST] - ${r(0.11)} per 1000\n` +
               `▫️ /6855 - telegram post views [1 post] [CHEAPEST] - ${r(0.09)} per 1000\n\n` +
               `👥 *TELEGRAM MEMBERS*\n` +
               `▫️ /7153 - Telegram Members [Refill 3 Days] - ${r(0.52)} per 1000\n` +
               `▫️ /6787 - Telegram Members [Mixed, Cheap] - ${r(0.38)} per 1000\n` +
               `▫️ /3274 - Telegram Channel Member [Mixed, Cheap] - ${r(0.52)} per 1000\n\n` +
               `🛒 Order karne ke liye code *type* karein (e.g. /5153)`;

    const keyboard = new InlineKeyboard().text("⬅️ Back", "main_services");
    await ctx.editMessageText(text, { reply_markup: keyboard, parse_mode: "Markdown" });
});

bot.callbackQuery("platform_instagram", async (ctx) => {
    const user = getOrCreateUser(ctx.from.id);
    const pref = user.currency;
    const r = (usd) => pref === "INR" ? `₹${(usd * USD_TO_INR_RATE).toFixed(2)}` : `$${usd.toFixed(2)}`;

    let text = `🔸 *INSTAGRAM SERVICES* [Currency: ${pref}]\n\n` +
               `❤️ *LIKES*\n` +
               `▫️ /7802 - Likes [Speed 20K/Hr] - ${r(0.21)}\n` +
               `▫️ /7526 - Likes [HQ Instant] - ${r(0.26)}\n` +
               `▫️ /7374 - Likes [Indian Mixed] - ${r(0.19)}\n\n` +
               `👥 *FOLLOWERS & VIEWS*\n` +
               `▫️ /3602 - Followers [30 Days Refill] - ${r(3.12)}\n` +
               `▫️ /1658 - Followers [Max 200K] - ${r(1.82)}\n` +
               `▫️ /1961 - Followers [Max 10K] - ${r(2.48)}\n` +
               `▫️ /8810 - Followers [No Refill] - ${r(1.77)}\n` +

File: bot.js
require('dotenv').config();
const { Bot, InlineKeyboard } = require('grammy');
const { run } = require('@grammyjs/runner');
const fs = require('fs');
const config = require('./config');
const m = require('./menuHandlers');
const o = require('./orderHandlers');
let SERVICES_MASTER_DATA = require('./services');

if (!config.BOT_TOKEN) process.exit(1);
const bot = new Bot(config.BOT_TOKEN);

const DB_FILE = './database.json';
let DYNAMIC_USER_DB = { users: {}, pending_deposits: {}, dynamic_config: {} };

if (fs.existsSync(DB_FILE)) {
    try {
        const parsed = JSON.parse(fs.readFileSync(DB_FILE, 'utf-8'));
        if (parsed.users && parsed.pending_deposits) DYNAMIC_USER_DB = parsed;
        else DYNAMIC_USER_DB = { users: parsed, pending_deposits: {}, dynamic_config: {} };
    } catch (e) { DYNAMIC_USER_DB = { users: {}, pending_deposits: {}, dynamic_config: {} }; }
}
if (!DYNAMIC_USER_DB.dynamic_config) DYNAMIC_USER_DB.dynamic_config = {};

function forceSaveDatabase() { fs.writeFileSync(DB_FILE, JSON.stringify(DYNAMIC_USER_DB, null, 4), 'utf-8'); }
function getLiveUpi() { return DYNAMIC_USER_DB.dynamic_config.upi_id || config.UPI_ID; }
function getLiveTrc() { return DYNAMIC_USER_DB.dynamic_config.usdt_trc20 || config.USDT_TRC20; }
function getLiveBep() { return DYNAMIC_USER_DB.dynamic_config.usdt_bep20 || config.USDT_BEP20; }

function getLocalUser(id, name = "User") {
    if (!DYNAMIC_USER_DB.users[id]) {
        DYNAMIC_USER_DB.users[id] = { username: name, balance_usd: 0.0, total_deposit_usd: 0.0, spent_usd: 0.0, orders_count: 0, cancelled_orders: 0, pending_orders: 0, history: [], currency: "INR", pending_service: null, pending_qty: null, pending_cost_usd: null, awaiting_custom_qty: false, awaiting_deposit_amt: false, chosen_pay_method: null, awaiting_utr: false, current_deposit_amt: 0, current_order_num: 0, chosen_network: "", is_banned: false }; forceSaveDatabase();
    } return DYNAMIC_USER_DB.users[id];
}
function formatMoneyLocal(usd, pref) { return pref === "INR" ? `₹${(usd * config.USD_TO_INR_RATE).toFixed(2)}` : `$${usd.toFixed(2)}`; }

bot.use(async (ctx, next) => {
    if (ctx.from && DYNAMIC_USER_DB.users[ctx.from.id]?.is_banned) {
        if (ctx.callbackQuery) return ctx.answerCallbackQuery({ text: "🛑 Aapko BAN kiya gaya hai!", show_alert: true });
        return ctx.reply("❌ *Aapko is bot se BAN kar diya gaya hai!*");
    } await next();
});

bot.command("start", m.start);
bot.callbackQuery("back_to_menu", m.backMenu);
bot.callbackQuery("check_balance", m.checkBalance);
bot.callbackQuery("my_profile", m.myProfile);
bot.callbackQuery("my_channels", m.myChannels);
bot.callbackQuery("main_promo", m.mainPromo);
bot.callbackQuery("main_support", m.mainSupport);
bot.callbackQuery("toggle_currency", m.toggleCurrency);
bot.callbackQuery("main_services", o.servicesMenu);
bot.callbackQuery("p_tg", o.tgMenu);
bot.callbackQuery("p_ig", o.igMenu);
bot.callbackQuery("p_fb", o.fbMenu);
bot.callbackQuery("p_yt", o.ytMenu);
bot.hears(/^\/\d+$/, o.handleSlashCode);
bot.callbackQuery(/^q_\d+_(.+)$/, o.handleQtyButtons);
bot.callbackQuery("main_orders", o.ordersHistory);

bot.callbackQuery("main_add_funds", async (ctx) => {
    const kb = new InlineKeyboard().text("Payment via UPI", "pay_via_upi").text("Payment via USDT", "pay_via_usdt").row().text("BACK", "back_to_menu");
    await ctx.editMessageText("💳 *Select Payment Method / पेमेंट का तरीका चुनें:*", { reply_markup: kb, parse_mode: "Markdown" });
});

bot.callbackQuery(/^pay_(via_upi|via_usdt)$/, async (ctx) => {
    const u = getLocalUser(ctx.from.id); u.chosen_pay_method = ctx.callbackQuery.data; u.awaiting_deposit_amt = true;
    if (u.chosen_pay_method === "pay_via_upi") { await ctx.editMessageText(`💰 *Enter Amount:* UPI\n\n¼कृपया वह राशि (INR ₹) टाइप करें जो आप जोड़ना चाहते हैं:\nPlease enter the amount (INR ₹) you want to add:`); }
    else { await ctx.editMessageText(`💰 *Enter Amount:* USDT\n\nकृपया वह राशि (USDT) टाइप करें जो आप जोड़ना चाहते हैं:\nPlease enter the amount (USDT) you want to add:`); }
});

bot.callbackQuery("user_complete_pay_via_upi", async (ctx) => {
    const u = getLocalUser(ctx.from.id); u.awaiting_utr = true; const orderNum = Math.floor(100000 + Math.random() * 900000); u.current_order_num = orderNum;
    await ctx.editMessageText("💵 *Payment Initiated!* ✅\n\n📊 *Expected Amount:* `₹" + u.current_deposit_amt.toFixed(2) + "`\n🆔 *Order Number:* `# " + orderNum + "`\n\n**⚠️ SUBMIT UTR TRANSACTION ID:**\nBhai, ab apna 12-digit UTR/Reference number niche message box mein type karke send karo aur sath mein payment ka screenshot bhi attach karke bhejo:", { parse_mode: "Markdown" });
});

bot.callbackQuery("usdtnet_bep20", async (ctx) => {
    const u = getLocalUser(ctx.from.id); u.chosen_network = "bep20"; const kb = new InlineKeyboard().text("CONFIRM PAYMENT", "usdt_confirm_click").row().text("BACK", "main_add_funds");
    await ctx.editMessageText("🪙 *USDT BEP20 MANUAL DEPOSIT*\n\n💵 *Amount to Pay:* $" + u.current_deposit_amt.toFixed(2) + "\n📍 *Address:* `" + getLiveBep() + "`\n\n👉 *Instructions:* Diye gaye Address par exactly $" + u.current_deposit_amt.toFixed(2) + " send karke neeche *CONFIRM PAYMENT* par click karein.", { reply_markup: kb, parse_mode: "Markdown" });
});

bot.callbackQuery("usdtnet_trc20", async (ctx) => {
    const u = getLocalUser(ctx.from.id); u.chosen_network = "trc20"; const kb = new InlineKeyboard().text("CONFIRM PAYMENT", "usdt_confirm_click").row().text("BACK", "main_add_funds");
    await ctx.editMessageText("🪙 *USDT TRC20 MANUAL DEPOSIT*\n\n💵 *Amount to Pay:* $" + u.current_deposit_amt.toFixed(2) + "\n📍 *Address:* `" + getLiveTrc() + "`\n\n👉 *Instructions:* Diye gaye Address par exactly $" + u.current_deposit_amt.toFixed(2) + " send karke neeche *CONFIRM PAYMENT* par click karein.", { reply_markup: kb, parse_mode: "Markdown" });
});

bot.callbackQuery("usdt_confirm_click", async (ctx) => {
    const u = getLocalUser(ctx.from.id); u.awaiting_utr = true; const orderNum = Math.floor(100000 + Math.random() * 900000); u.current_order_num = orderNum;
    await ctx.editMessageText("🪙 *USDT Deposit Initiated!* ✅\n\n📊 *Requested Amount:* `$" + u.current_deposit_amt.toFixed(2) + "`\n🌐 *Network:* `" + u.chosen_network.toUpperCase() + "`\n🆔 *Order Number:* `# " + orderNum + "`\n\n**⚠️ SUBMIT TRANSACTION ID / HASH:**\nBhai, apni USDT Transaction Hash ID niche message box mein type karke send karo:", { parse_mode: "Markdown" });
});

// 👑 Galti No. 1 Fixed: Dynamic indices mapping updated strictly for approval routines
bot.callbackQuery(/^adm_(acc|can)_(.+)_(.+)$/, async (ctx) => {
    if (ctx.from.id !== config.ADMIN_ID) return;
    const parts = ctx.callbackQuery.data.split("_"), action = parts[1], userId = parseInt(parts[2]), refKey = parts[3];
    const depositData = DYNAMIC_USER_DB.pending_deposits[refKey]; if (!depositData) return ctx.answerCallbackQuery({ text: "❌ Link Expired or Already Approved!", show_alert: true });
    const u = getLocalUser(userId);
    if (action === "acc") {
        u.balance_usd += depositData.amount_usd; u.total_deposit_usd += depositData.amount_usd; forceSaveDatabase();
        await bot.api.sendMessage(userId, `✅ *Payment Added Successful!* 💰\n\nBhai tumhara payment verify ho gaya hai.\n✨ *Added Amount:* ${formatMoneyLocal(depositData.amount_usd, u.currency)}\n💳 *Total Balance:* ${formatMoneyLocal(u.balance_usd, u.currency)}`, { parse_mode: "Markdown" });
        await ctx.editMessageText(`✅ Request Accepted for User ${userId}`);
    } else {
        await bot.api.sendMessage(userId, `❌ *Payment Request Cancelled!*`); await ctx.editMessageText(`❌ Cancelled for User ${userId}`);
    } delete DYNAMIC_USER_DB.pending_deposits[refKey]; forceSaveDatabase();
});

bot.command("admin", async (ctx) => {
    if (ctx.from.id !== config.ADMIN_ID) return;
    await ctx.reply(`⚙️ *HAPPY REACTION Super Admin Control Panel*\n\n➕ *User Matrix Controls:*\n▫️ \`/ban USER_ID\` \n▫️ \`/unban USER_ID\` \n▫️ \`/addbalance USER_ID AMOUNT\` \n▫️ \`/deductbalance USER_ID AMOUNT\` \n▫️ \`/checkuser USER_ID\` -> View total profile & order metrics\n\n🛠️ *SMM Live Controls:*\n▫️ \`/addservice ID Rate Type Name\`\n▫️ \`/updateservice ID NewRate\`\n▫️ \`/delservice ID\`\n\n💳 *Payment Credentials Controls (Live Change):*\n▫️ \`/setupi NEW_UPI_ID\` \n▫️ \`/settrc20 WALLET_ADDRESS\` \n▫️ \`/setbep20 WALLET_ADDRESS\``, { parse_mode: "Markdown" });
});

bot.command("ban", async (ctx) => {
    if (ctx.from.id !== config.ADMIN_ID) return;
    const target = parseInt(ctx.message.text.split(" ")[1]); if (isNaN(target) || !DYNAMIC_USER_DB.users[target]) return ctx.reply("❌ ID nahi mili!");
    DYNAMIC_USER_DB.users[target].is_banned = true; forceSaveDatabase(); await ctx.reply(`🚫 User \`${target}\` BAN ho gaya hai.`);
});

bot.command("unban", async (ctx) => {
    if (ctx.from.id !== config.ADMIN_ID) return;
    const target = parseInt(ctx.message.text.split(" ")[1]); if (isNaN(target) || !DYNAMIC_USER_DB.users[target]) return ctx.reply("❌ ID nahi mili!");
    DYNAMIC_USER_DB.users[target].is_banned = false; forceSaveDatabase(); await ctx.reply(`✅ User \`${target}\` UNBAN ho gaya hai.`);
});

// 👑 Galti No. 2 Fixed: Template literal string concatenated properly
bot.command("addbalance", async (ctx) => {
    if (ctx.from.id !== config.ADMIN_ID) return;
    const args = ctx.message.text.split(" "), target = parseInt(args[1]), amt = parseFloat(args[2]);
    if (isNaN(target) || isNaN(amt) || !DYNAMIC_USER_DB.users[target]) return ctx.reply("❌ Format: \`/addbalance USER_ID AMOUNT\`");
    DYNAMIC_USER_DB.users[target].balance_usd += amt; forceSaveDatabase(); 
    await ctx.reply("💰 *Successfully Added $*" + amt.toFixed(2) + " *to User:* `" + target + "`");
    try { await bot.api.sendMessage(target, "✨ *Admin dwara tumhare account mein $*" + amt.toFixed(2) + " *add kar diye gaye hain!*"); } catch(e) {}
});

bot.command("deductbalance", async (ctx) => {
